service/controller/release/resource/status/cluster.go

```diff
@@ -0,0 +1,205 @@
+package status
+
+import (
+	"context"
+
+	apiexlabels "github.com/giantswarm/apiextensions/v2/pkg/label"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	azurecapi "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+
+	"github.com/giantswarm/release-operator/service/controller/key"
+)
+
+type TenantCluster struct {
+	ID               string
+	OperatorVersion  string
+	ProviderOperator string
+	ReleaseVersion   string
+}
+
+// Takes a list of tenant clusters and returns two maps containing the versions of their release and operator versions.
+func consolidateClusterVersions(clusters []TenantCluster) (map[string]bool, map[string]map[string]bool) {
+	releaseVersions := make(map[string]bool)
+	operatorVersions := make(map[string]map[string]bool)
+
+	// operatorVersions is a nested map including the operator name and version
+	// e.g. operatorVersions["aws-operator"]["8.7.6"]:true
+
+	for _, c := range clusters {
+		releaseVersions[c.ReleaseVersion] = true
+
+		if operatorVersions[c.ProviderOperator] == nil {
+			operatorVersions[c.ProviderOperator] = make(map[string]bool)
+		}
+		operatorVersions[c.ProviderOperator][c.OperatorVersion] = true
+	}
+
+	return releaseVersions, operatorVersions
+}
+
+// Returns a list of tenant clusters currently running on the installation.
+func (r *Resource) getCurrentTenantClusters(ctx context.Context) ([]TenantCluster, error) {
+
+	var tenantClusters []TenantCluster
+	{
+		awsClusters, err := r.getCurrentAWSClusters(ctx)
```

Comment:
Since `getCurrentAWSClusters`, `getCurrentAzureClusters`, `getLegacyAWS/KVM/AzureClusters`,  all have the same signature, you could loop over them and append them in sequence to cut down on some code.

Reply:
Haha I wasn't sure if that was good Go style, it felt too pythonic -- I like that a lot more

Reply:
You tend to be limited in Go due to typing, but since these all return the same type, I don't see a problem with it. The main goal is clarity/readability and I think that is maintained. In the future we can drop the legacy CRs just by removing elements from the slice which is nice.